rustworkx-core/src/dag_algo.rs

```diff
@@ -314,6 +319,176 @@ where
     Ok(Some((path, path_weight)))
 }
 
+/// Return a list of graph layers
+///
+/// A layer is a subgraph whose nodes are disjoint, i.e.,
+/// a layer has depth 1. The layers are constructed using a greedy algorithm.
+///
+/// Arguments:
+///
+/// * `graph` - The graph to get the layers from
+/// * `first_layer` - A list of node ids for the first layer. This
+///     will be the first layer in the output
+///
+/// ```
+/// use rustworkx_core::petgraph::prelude::*;
+/// use rustworkx_core::dag_algo::layers;
+/// use rustworkx_core::dictmap::*;
+///
+/// let edge_list = vec![
+///  (0, 1),
+///  (1, 2),
+///  (2, 3),
+///  (3, 4),
+/// ];
+///
+/// let graph = DiGraph::<u32, u32>::from_edges(&edge_list);
+/// let layers: Vec<Vec<NodeIndex>> = layers(&graph, vec![0.into(),1.into()]).unwrap().collect();
+/// let expected_layers: Vec<Vec<NodeIndex>> = vec![
+///     vec![0.into(),1.into()],
+///     vec![1.into(),2.into()],
+///     vec![2.into(),3.into()],
+///     vec![3.into(),4.into()],
+///     vec![4.into()]
+/// ];
+/// assert_eq!(layers, expected_layers)
+/// ```
+pub fn layers<G>(
+    graph: G,
+    first_layer: Vec<G::NodeId>,
+) -> Result<LayersIter<G, G::NodeId>, LayersError>
+where
+    G: NodeIndexable // Used in from_index and to_index.
+        + NodeCount // Used in find_cycle
+        + EdgeCount // Used in find_cycle
+        + Visitable // Used in find_cycle
+        + IntoNodeIdentifiers // Used for .node_identifiers
+        + IntoNeighborsDirected // Used for .neighbors_directed
+        + IntoEdgesDirected, // Used for .edged_directed
+    <G as GraphBase>::NodeId: Debug + Copy + Eq + Hash,
+{
+    let cur_layer = first_layer;
+    // Throw error if a cycle is found at the current node
+    if let Some(node) = &cur_layer.first() {
+        let check_cycle = find_cycle(&graph, Some(**node));
+        if !check_cycle.is_empty() {
+            return Err(LayersError(Some(format!(
+                "Found cycle at {:?}",
+                check_cycle
+            ))));
+        }
+    }
```

Comment:
I'm still worried about this, you're adding extra overhead for a complete DFS of the graph just for error checking when it didn't exist before. This could be a noticeable performance regression especially for large graphs. If you think we need to do result handling here we should do the checking during the iteration since we're already iterating over the graph.

Reply:
I was thinking about it and the method you suggested before about having a set to keep track of all the nodes was right, but didn't seem to be working well for me, due to a programming error on my end. In previous iterations of me trying to fix that, I forgot to add the elements of the first layer into that set, which resulted in the traversal continuing infinitely. This should be fixed in a future commit.